sklearn/naive_bayes.py

```diff
@@ -117,6 +117,229 @@ def predict_proba(self, X):
+            well, to indicate to drop the columns or to pass them through
+            untransformed, respectively.
+        column(s) : string or int, array-like of string or int, slice, \
+boolean mask array or callable
+            Indexes the data on its second axis. Integers are interpreted as
+            positional columns, while strings can reference DataFrame columns
+            by name.  A scalar string or int should be used where
+            ``transformer`` expects X to be a 1d array-like (vector),
+            otherwise a 2d array will be passed to the transformer.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.naive_bayes import GeneralNB, GaussianNB, BernoulliNB
+    >>> X = np.array([[1.5, 2.3, 5.7, 0, 1],
+    ...               [2.7, 3.8, 2.3, 1, 0],
+    ...               [1.7, 0.1, 4.5, 1, 0]])
+    >>> y = np.array([1, 0, 0])
+    >>> clf = GeneralNB([
+    >>>     ("gaussian", GaussianNB(), [0, 1, 2]),
+    >>>     ("bernoulli", BernoulliNB(), [3, 4])
+    >>> ])
+    >>> clf.fit(X, y)
+    >>> print(clf.predict([[1.5, 2.3, 5.7, 0, 1]]))
+    [1]
+    >>> print(clf.score([[2.7, 3.8, 1, 0, 1]],[0]))
+    1.0
+    """
+    # TODO check_is_fitted
+    # TODO consider jll for each estimator
+
+    def __init__(self, distributions):
+        self.distributions = distributions
+        self._fits = []
+        self._is_fitted = False
+
+    @property
+    def distributions_(self):
+        return [(name, distr) for name, distr, _ in self.distributions]
+
+    @distributions_.setter
+    def distributions_(self, value):
+        self.distributions = [
+            (name, distr, col) for ((name, distr), (_, _, col))
+            in zip(value, self.distributions)]
+
+    def get_params(self, deep=True):
+        return self._get_params('distributions_', deep=deep)
+
+    def set_params(self, **kwargs):
+        self._set_params('distributions_', **kwargs)
+        return self
+
+    def fit(self, X, y):
+        """Fit Gaussian Naive Bayes according to X, y
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training vectors, where n_samples is the number of samples
+            and n_features is the number of features.
+        y : array-like, shape (n_samples,)
+            Target values.
+        distributions : list of tuples
+            A list of (NB, features) tuples, where NB is 'BernoulliNB',
+            'GaussianNB', 'MultinomialNB', 'ComplementNB' or 'CategoricalNB',
+            and features is a list of indices.
+
+        Returns
+        -------
+        self : object
+        """
+        self._validate_distributions(X)
+        X, y = check_X_y(X, y)
+        y = column_or_1d(y, warn=True)
+
+        # Should be the same after the validation
+        self.classes_ = np.unique(y)
+
+        inits = [(nb, features) for (_, nb, features) in self.distributions]
+
+        self._fits = [(nb.fit(X[:, features], y), features)
```

Comment:
why don't we just set `distributions_` directly here, with the fitted estimators?

Reply:
And `X[:, features]` won't work for several formats of `Features`. Please check how ColumnTransformer handles its column selection.

Note that in ColumnTransformer we also handle named columns, exploiting this feature of pd.DataFrames. Here we have already converted X to a numpy array.

Reply:
Alright, I've added support for pd.DataFrames